### PR TITLE
[5.5] [Security] Close `remember_me` Timing Attack Vector

### PR DESCRIPTION
**Pull request by mcordingley**
The current `remember_me` token verification process leaves the application open to a [timing attack](http://blog.ircmaxell.com/2014/11/its-all-about-time.html).

Since the default is for the token to be stored as a cookie and for cookies to be encrypted, an attacker would have to know the application secret to exploit this. However, should a custom guard be used or cookies not be encrypted, it becomes possible to tease this value out.

The proposed change switches to comparing the token using a constant-time comparison. This makes it impossible to learn the value of the token by timing responses, independently of guard or encryption settings.

### CODE REVIEW

File: src/Illuminate/Auth/DatabaseUserProvider.php [hunk 1]
```diff
@@ -70,10 +70,9 @@ public function retrieveByToken($identifier, $token)
     {
         $user = $this->conn->table($this->table)
```
Comment (GrahamCampbell) [inline src/Illuminate/Auth/DatabaseUserProvider.php]: This is surely the same as `$user = $this->conn->table($this->table)->find($identifier);` now?

Comment (richardkeep) [reply src/Illuminate/Auth/DatabaseUserProvider.php]: Yes. Definitely

**Comment by Naxon** [comment]
After upgrading I get this error:
`hash_equals(): Expected known_string to be a string, null given (View: /home/vagrant/Code/the-gate/resources/views/welcome.blade.php)`

In this line:
`return $model && hash_equals($model->getRememberToken(), $token) ? $model : null;`

**Comment by crynobone** [comment]
A recent PR change the way remember token updated on logout, now during logout it will reset to `null` which would cause the issue

**Comment by Naxon** [comment]
#21328

**Comment by themsaid** [comment]
@crynobone can you please link me to this change? it seems like the remember_me token is renewed on logout, but never set to null!

**Comment by at-phucnguyen** [comment]
@taylorotwell 
How about previous versions like 5.2?  was it fixed?

**Comment by mcordingley** [comment]
@at-phucnguyen I don't think 5.2 is getting any more updates, even security ones. If you're worried about security, upgrade as soon as you're able. You're missing out on far more than just this one issue.

**Comment by bytestream** [comment]
A similar fix should be applied to 5.1 given it's LTS for security fixes exists until June 2018?

**Comment by Quix0r** [comment]
Has the (possible, not reproduced by me) introduced bug reported by @Naxon being addressed?

**Comment by Quix0r** [comment]
Is #21323 the fix?

**Comment by cerw** [comment]
How about 5.4?

**Comment by devcircus** [comment]
Security updates stopped in January for 5.4. 
https://laravel.com/docs/5.6/releases#support-policy

**Pull request merged**